File: search_sort/heapsort.c
#include <stdio.h>
#define LEN 20 
#define MAX 100

void swap(int *a, int *b){
	int tmp = *a;
	*a = *b;
	*b = tmp;
}

void heapify(int A[], int idx, int max){
	int largest = idx;
	int left = idx*2+1;
	int right = idx*2+2;
	if(left<max && A[left]>A[idx]) largest = left;
	if(right<max && A[right]>A[largest]) largest = right;
	if(largest != idx){
		swap(&A[largest], &A[idx]);
		heapify(A, largest, max);
	}
}

void build_maxheap(int A[], int len){
	int i;
	for(i=len/2-1; i>=0; i--)
		heapify(A, i, len);
}


void heapsort(int A[], int len){
	build_maxheap(A, len);
	int i;
	for(i=len-1; i>0; i--){
//		int j;
//		for(j=0; j<len; j++) printf("%d ", A[j]); printf("\n");
		swap(&A[0], &A[i]);
		heapify(A, 0, i);
	}
}	

int main(){
	int i;
	int test[LEN];
	for(i=0; i<LEN; i++) test[i] = rand()%MAX; 
	for(i=0; i<LEN; i++) printf("%d ", test[i]); 
	printf("\n");

	heapsort(test, LEN);

	for(i=0; i<LEN; i++) printf("%d ", test[i]); 
	printf("\n");
}

File: python/heapSort.py
def heapify(A, idx, end):
    left = idx*2 + 1
    right = idx*2 + 2
    maxIdx = idx
    if left < end and A[left] > A[idx]:
        maxIdx = left
    if right < end and A[right] > A[maxIdx]:
        maxIdx = right
    if maxIdx != idx:
        A[idx],A[maxIdx] = A[maxIdx],A[idx]
        heapify(A, maxIdx, end)

def buildHeap(A):
    n = len(A)
    for i in range(n//2-1, -1, -1):
        heapify(A, i, n)
        # print A

def heapSort(A):
    buildHeap(A)
    n = len(A)
    for i in range(0, n):
        A[0], A[n-i-1] = A[n-i-1], A[0]
        heapify(A, 0, n-i-1)

heap = [3,1,6,7,9,2,4,8,5]
buildHeap(heap)
print(heap)

h = 0
while (1<<h) <= len(heap):
    print(heap[(1<<h)-1:(1<<(h+1))-1])
    h += 1

heapSort(heap)
print(heap)
